src/openfermion/utils/_bch_expansion.py

```diff
@@ -0,0 +1,157 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""Module to efficiently compute the Baker-Campbell-Hausdorff formula."""
+
+import itertools
+import numpy as np
+from scipy.misc import comb, factorial
+
+from openfermion.utils import commutator
+
+def bch_expand(x, y, order):
+    """Compute log[e^x e^y] using the Baker-Campbell-Hausdorff formula
+    Args:
+        x: An operator for which multiplication and addition are supported.
+            For instance, a QubitOperator, FermionOperator or numpy array.
+        y: The same type as x.
+        order(int): The order to truncate the BCH expansions.
+
+    Returns:
+        z: The truncated BCH operator.
+
+    Raises:
+        ValueError: operator x is not same type as operator y.
+        ValueError: invalid order parameter.
+        ValueError: order exceeds maximum order supported.
+    """
+
+    max_order = 11
```

Comment:
I think because the implementation is still an exponential complexity (though much better than naive) it's still necessary to set an upper bound, since we don't want to blow up the user's RAM, we use mostly on the fly computation and this can lead to super long wait time. I think it's reasonable that we simply raise an exception but perhaps with better explanation.

Reply:
I agree that setting an maximum order is not ideal. I will argue that a hard cutoff doesn't achieve anything: either computing the 11th order is already too slow (and the cutoff doesn't achieve its intended purpose), or computing the 12th order still only takes a reasonable time (and you are blocking the user from doing reasonable work).